_episodes/07-web-app.md

```diff
@@ -0,0 +1,234 @@
+~~~
+      Emma  Jane
+Time
+0        0     0
+1       10    15
+2       20    20
+3       30    25
+~~~
+{: .output}
+
+If we call the `plot` function (method) directly on the `pandas` dataframe, we
+get a nice default style:
+
+~~~
+table.plot();
+~~~
+{: .language-python}
+
+![Default Dataframe Plot](../fig/07_default_df_plot.png)
+
+Indeed, we get a legend and an `xlabel`... for free! Here, we can see the
+value of working with *labelled* data.
+
+If we wanted to plot the position of a single runner, say, Jane, we would do
+(see previous episode):
+
+~~~
+table['Jane'].plot();
+~~~
+{: .language-python}
+
+![Default Timeseries Plot](../fig/07_default_ts_plot.png)
+
+If we wanted to plot the positions only up to a certain time point, say, up to
+the third time point, we would do:
+
+~~~
+table[:3].plot();
+~~~
+{: .language-python}
+
+![Plot for a Subset of Rows](../fig/07_subset_df_plot.png)
+
+## Selecting and Plotting them Interactively
+
+[`ipywidgets`](https://ipywidgets.readthedocs.io/) is a library for exploring
+code and data interactively. We import it as follows:
+
+~~~
+import ipywidgets as widgets
+~~~
+{: .language-python}
+
+It provides a function, named `interact`, which allows us to create basic user
+interface (UI) controls. For example, we may use a dropdown menu for
+selecting the runner (Emma or Jane) whose position we want to visualize.
+Likewise, we may use a slider for selecting the end value of the time points.
+
+To begin with, we need to write a function where the parameter is the variable
+we want to control (interact with). In the first case, this could be simply:
+
+~~~
+def f(x):
+    table[x].plot();
+~~~
+{: .language-python}
+
+Then, we need to call `interact` with this function as the first argument and
+the variable to control as the second argument:
+
+~~~
+widgets.interact(f, x={'Emma', 'Jane'});
+~~~
+{: .language-python}
+
+![Screenshot of Interactive Plot with Dropdown](../fig/07_dropdown_screenshot.png)
+
+> ## Choosing Descriptive Function and Variable Names
+>
+> 1. Write a function to plot the timeseries up to a given time point.
+> 2. What do you think about the names of your function and its parameter?
+>
+> > ## Solution
+> >
+> > 1. To plot the timeseries (which are the columns of dataframe `table`) up
+> > to time point `xx`, we can use the following function:
+> > ~~~
+> > def ff(xx):
+> >     table[:xx].plot();
+> > ~~~
+> > {: .language-python}
+> >
+> > Here we could have a discussion about whether `xx` is included or not: If
+> > `xx` is a row index, it is excluded (as we have seen in episode 02 on
+> > lists). If `xx` is an actual time value (timestamp), then it is included.
+> >
+> > 2. We have named this function `ff` because we already had a function
+> > named `f` (this habit comes from a sub-culture in the R community). To
+> > make the code easier to understand and maintain, it would be better to
+> > name it, say, `truncate_timeseries`; likewise, the `xx` parameter could be
```

Comment:
I don't really see a purpose of talking about R community here, unless there is any specific story (personal or community related). I would just name the functions in a better way from the beginning

Reply:
As I replied above, the 'not-so-good way' is a pretext for a challenge, for students to practise the better way. :slightly_smiling_face: 

The (real) story is that `pandas` and dataframes in Python *come straight from R* so it's good to know (although, of course, one can absolutely use `pandas` without ever knowing about this...). When I started doing data science (about 10 years ago), I was using mostly R, `pandas` didn't exist... So, yes, I could share a personal anecdote as well. :wink: